README.rst
plone.recipe.zope2zeoserver
===========================

This recipe creates and configures a Zope 2 ZEO server in parts. It also
installs a control script, which is like zeoctl, in the bin/ directory.
The name of the control script is the the name of the part in buildout.

You can use it with a part like this::

  [zeoserver]
  recipe = plone.recipe.zope2zeoserver
  zope2-location = /path/to/zope2/install
  zeo-address = 8100

.. ATTENTION::
  This recipe works with Zope 2 versions prior to Zope 2.12 or ZODB3
  versions prior to ZODB 3.8. If you want to use newer versions of any
  of the two (for example for Plone 4) please use the
  `plone.recipe.zeoserver`_: recipe instead.

.. _`plone.recipe.zeoserver`: http://pypi.python.org/pypi/plone.recipe.zeoserver


Options
-------

zope2-location
  The path where Zope 2 is installed. If you are also using the
  plone.recipe.zope2install recipe, and you have that configured as a part
  called 'zope2' prior to the zope2zeoserver part, you can use
  ${zope2:location} for this parameter. You must ensure the zope2zeoserver part
  is run *after* the zope2install one.

zeopack
  The path to the zeopack.py backup script. A wrapper for this will be
  generated in bin/zeopack, which sets up the appropriate environment to
  run this. Defaults to "${zope2-location}/utilities/ZODBTools/zeopack.py".
  Set this option to an empty value if you do not want this script to be
  generated.

repozo
  The path to the repozo.py backup script. A wrapper for this will be
  generated in bin/repozo, which sets up the appropriate environment for
  running this. Defaults to "${zope2-location}/utilities/ZODBTools/repozo.py".
  Set this to an empty value if you do not want this script to be generated.

zeo-conf
  A relative or absolute path to a zeo.conf file. If this is not given, a
  zeo.conf will be generated based on the the options below.

The following options all affect the generated zope.conf.

zeo-address
  Give a port for the ZEO server (either specify the port number only (with
  'localhost' as default) or you use the format ``host:port``). Defaults to 8100.

effective-user
  The name of the effective user for the ZEO process. Defaults to not setting
  an effective user.

invalidation-queue-size
  The invalidation-queue-size used for the ZEO server. Defaults to 100.

zeo-log
  The filename of the ZEO log file. Defaults to var/log/${partname}.log

zeo-log-format
  Format of logfile entries. Defaults to %(asctime)s %(message)s

zeo-log-custom
  A custom section for the eventlog, to be able to use another
  event logger than `logfile`. `zeo-log` is still used to set the logfile
  value in the runner section.

storage-number
  The number used to identify a storage.

file-storage
  The filename where the ZODB data file will be stored.
  Defaults to var/filestorage/Data.fs.

blob-storage
  The folder where the ZODB blob data files will be stored.

socket-name
  The filename where ZEO will write its socket file.
  Defaults to var/zeo.zdsock.

authentication-database
  The filename for a authentication database. Only accounts listed in this
  database will be allowed to access the ZEO server.

  The format of the database file is::

    realm <realm>
    <username>:<hash>

  Where the hash is generated via::

    import sha
    string = "%s:%s:%s" % (username, realm, password)
    sha.new(string).hexdigest()

authentication-realm
  The authentication realm. Defaults to 'ZEO'

pack-days
  Specify of days for the zeopack script to retain of history. Defaults to
  one day.

pack-user
  If the ZEO server uses authentication, this is the username used by the
  zeopack script to connect to the ZEO server.

pack-password
  If the ZEO server uses authentication, this is the password used by the
  zeopack script to connect to the ZEO server.

zeo-conf-additional
  Give additional lines to zeo.conf. Make sure you indent any lines after
  the one with the parameter.

monitor-address
  The address at which the monitor server should listen.
  The monitor server provides server statistics in a simple text format.

relative-paths
  Set this to `true` to make the generated scripts use relative
  paths. You can also enable this in the `[buildout]` section.

eggs
  Set if you need to include other packages as eggs e.g. for making
  application code available on the ZEO server side for performing
  conflict resolution (through the _p_resolveConflict() handler).

read_only
  Set if you want to put the ZEO server in read-only mode.
  No modifications will be made on the Data.fs.

Reporting bugs or asking questions
----------------------------------

We have a shared bugtracker and help desk on Launchpad:
https://bugs.launchpad.net/collective.buildout/


CHANGES.txt
Change History
==============

1.6 (unreleased)
----------------

- Nothing changed yet.


1.5 (2013-01-29)
----------------

- LP #503686: Added support for setting ZEO log level.
  [baijum]

- Added note on 'eggs' option.
  [ajung]

- Added option to customize repozo and zeopack names
  [do3cc]

1.4 (2009-10-15)
----------------

* Make sure that it's still possible to specify a storage to pack using the -S
  parameter for bin/zeopack.
  [davisagli]

* Documented the format of the authentication database file.
  [hannosch]

* Added support for more than one zeoinstance per buildout on win32 environments
  [erico_andrei]

* Added support for relative-paths in the script generation.
  [jvloothuis]

1.3 (2009-05-14)
----------------

* Added the new `pack-days`, `pack-user` and `pack-password` options to give
  more flexible options for the zeopack script and support packing databases
  which require authentication. Also reworked the zeopack internals to support
  packing blob storages. The default for packing is now to retain one day of
  database history.
  [hannosch]

1.2 (2009-02-25)
----------------

* Generate a bin/repozo script to perform backups using repozo.py (and
  set up the appropriate pythonpath for this to work).
  Copied from plone.recipe.zope2instance.
  [fschulze]

* Added support for the zeo.conf 'monitor-address' parameter.
  [mattss]

1.1 (2008-12-05)
----------------

* Pass PYTHONPATH generated by recipe into zopeservice.py, so that it
  is properly setup and passed to service controlled environment. This
  makes it possible to use third-party ZConfig extensions with ZEO,
  like iw.rotatezlogs.
  [sidnei]

1.0 (2008-11-04)
----------------

* Bump version number to 1.0 since this has been stable for a long time.
  [wichert]

* Add MANIFEST.in file so we can create a release from an export.
  [wichert]


0.17 (2008-11-04)
-----------------

- Make the log format configurable and change the default to include the
  current time.
  [wichert]


0.16 (2008-10-15)
-----------------

- Adjusted the Supervisord support to ignore the command sent to the script.
  This avoids a `positional arguments are not supported` error.
  [hannosch]

- The launch script now detects if it is being executed from Supervisord.
  This makes it more manageable for Supervisord since it no longer forks.
  [jvloothuis]


0.15 (2008-08-27)
-----------------

- Renamed zeoservice.py to zeoservice.py.in to avoid errors during egg
  creation, as it tries to compile the template file.
  [hannosch]


0.14 (2008-08-21)
-----------------

- Fix the test to not rely on a specific installation folder name.
  [dataflake]

- Added support for a unix socket path in `zeo-address` and expanded the
  test coverage for the `zeopack` script generation.
  [dataflake]

- Add support for ZEO authentication.
  [wichert]


0.13 (2008-03-27)
-----------------

- Fixed call to HandleCommandLine by computing the serviceClassString
  ourselves if our module name is not __main__ (ie, if it was imported
  from another module). This works around the fact that
  HandleCommandLine calls GetServiceClassString which in turn only
  calculates the full path if the original module is __main__.
  [sidnei]

- Fixed generation of runzeo.bat when using ZEO from egg, where ZConfig is not
  in ZODB3_HOME.
  [rochael]

- Only generate the Windows service scripts if there is a zope2-location, since
  it's currently the only place where the nt_svcutils can be found.
  [rochael]

- Fixed test breakages on Windows.
  [rochael]


0.12 (2008-03-02)
-----------------

- Fixed zeopack generation so it includes zope-location.
  [tarek]


0.11 (2008-02-19)
-----------------

- Made Python 2.3 compatible
  [fschulze]

- Added win32 specific scripts.
  [tarek]

- Added a `zeo-log-custom` option.
  [tarek]

- Added a test environement and a basic doctest to validate the recipe.
  [tarek]


0.10 (2008-01-09)
-----------------

- Calculate the package working set outside of the recipe constructor. Closes
  https://bugs.launchpad.net/collective.buildout/+bug/180084
  [rochael]


0.9 (2007-12-30)
----------------

- Added `invalidation-queue-size` and `storage-number` options.
  [hannosch]

- Merged rochael-blobsupport branch. Added support for ZODB 3.8 blob storage
  configuration. This adds the blob-storage option. This closes
  https://bugs.launchpad.net/collective.buildout/+bug/179115 and
  https://bugs.launchpad.net/collective.buildout/+bug/179117.
  [rochael, hannosch]

- Renamed `zope-conf-additional` option to `zeo-conf-additional`.
  [rochael]


0.8
---

- Small documentation update. Added link to the bugtracker.
  [hannosch]

- Have PID file's location default to `${buildout:directory}/var/${name}.pid`.
  Keeping the PID file in $INSTANCE_HOME gives trouble when buildout rebuilds
  the part.
  [nouri, mbenali]


0.7
---

- Change default location of zeo socket to var.
  [smcmahon]

- Add option to set socket-name.
  [smcmahon]


0.6
---

- Move the documentation to a separate README.txt and read that from setup.py.
  [wichert]

- Add a wrapper around zeopack which inserts default options for our ZEO server.
  [wichert]


0.5
---

- Added an option to set the effective-user.
  [optilude]


0.3
---

- Updated setup.py.
  [optilude]


0.2
---

- Similarly, there are some hardcoded dependencies on $INSTANCE_HOME/etc/.
  [optilude]


src/plone/recipe/zope2zeoserver/tests/test_docs.py
# -*- coding: utf-8 -*-
"""
Grabs the doctests in /tests
"""
__docformat__ = 'restructuredtext'

import unittest
import doctest
import sys
import re
import os
import shutil
import popen2
import StringIO

from zope.testing import doctest, renormalizing
import zc.buildout.testing, zc.buildout.easy_install

current_dir = os.path.abspath(os.path.dirname(__file__))
recipe_location = current_dir

for i in range(5):
    recipe_location = os.path.split(recipe_location)[0]

def doc_suite(test_dir, globs=None):
    """Returns a test suite, based on doctests found in /doctest."""
    suite = []
    if globs is None:
        globs = globals()

    flags = (doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE |
             doctest.REPORT_ONLY_FIRST_FAILURE)

    doctest_dir = test_dir

    # filtering files on extension
    docs = [os.path.join(doctest_dir, doc) for doc in
            os.listdir(doctest_dir) if doc.endswith('.txt')]

    for test in docs:
        suite.append(doctest.DocFileSuite(test, optionflags=flags, 
                                          globs=globs, 
                    setUp=zc.buildout.testing.buildoutSetUp,
                    tearDown=zc.buildout.testing.buildoutTearDown,
                    checker=renormalizing.RENormalizing([
                        zc.buildout.testing.normalize_path,
                        (re.compile(r'\S+buildout.py'), 'buildout.py'),
                        (re.compile(r'line \d+'), 'line NNN'),
                        (re.compile(r'py\(\d+\)'), 'py(NNN)'),
                        ])
                        ,
                      module_relative=False))

    return unittest.TestSuite(suite)

def test_suite():
    """returns the test suite"""
    return doc_suite(current_dir)

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')



src/plone/recipe/zope2zeoserver/ctl.py
##############################################################################
#
# Copyright (c) 2006-2007 Zope Corporation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""zeocltl -- control a ZEO server using zdaemon.

Usage: zeocltl [options] [action [arguments]]

Options:
-h/--help -- print usage message and exit
-i/--interactive -- start an interactive shell after executing commands
action [arguments] -- see below

Actions are commands like "start", "stop" and "status". If -i is
specified or no action is specified on the command line, a "shell"
interpreting actions typed interactively is started (unless the
configuration option default_to_interactive is set to false). Use the
action "help" to find out about available actions.
"""

import sys
import os

from ZEO import zeoctl
from ZEO import runzeo


if sys.platform[:3].lower() == "win":
    print 'For win32 platforms, runzeo.bat or zeoservice.exe should be used'
    print '%s is based on zdaemon, which is Linux specific' % sys.argv[0]
    print 'Aborting...'
    sys.exit(0)

def main(args=None):
    # When we detect Supervisord we need to make sure we do not fork a
    # sub process since Supervisord does not like that
    if 'SUPERVISOR_ENABLED' in os.environ:
        runzeo.main(args)
    else:
        zeoctl.main(args)


src/plone/recipe/zope2zeoserver/__init__.py
##############################################################################
#
# Copyright (c) 2006-2007 Zope Corporation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################

import sys, os, shutil
import zc.buildout
import zc.recipe.egg
import logging

join = os.path.join

curdir = os.path.dirname(__file__)

class Recipe:

    def __init__(self, buildout, name, options):
        self.egg = zc.recipe.egg.Egg(buildout, options['recipe'], options)
        self.buildout, self.options, self.name = buildout, options, name

        options['location'] = os.path.join(
            buildout['buildout']['parts-directory'],
            self.name,
            )
        options['bin-directory'] = buildout['buildout']['bin-directory']
        options['scripts'] = '' # suppress script generation.

        _, self.zodb_ws = self.egg.working_set()

        # Relative path support for the generated scripts
        relative_paths = options.get(
            'relative-paths',
            buildout['buildout'].get('relative-paths', 'false')
            )
        if relative_paths == 'true':
            options['buildout-directory'] = buildout['buildout']['directory']
            self._relative_paths = options['buildout-directory']
        else:
            self._relative_paths = ''
            assert relative_paths == 'false'

    _ws_locations = None

    def ws_locations(self):
        if self._ws_locations is None:
            self._ws_locations = [d.location for d in self.zodb_ws]
            # account for zope2-location, if provided
            zope2_location = self.options.get('zope2-location')
            if zope2_location is not None:
                software_home = os.path.join(zope2_location,
                                             'lib', 'python')
                assert os.path.isdir(software_home), (invalid_z2_location_msg %
                                                      zope2_location)
                self._ws_locations.append(software_home)
        return self._ws_locations
    ws_locations = property(ws_locations)

    def install(self):
        options = self.options
        location = options['location']

        # evil hack alert!
        # hopefully place ZODB in sys.path
        orig_sys_path = tuple(sys.path)
        sys.path[0:0] = self.ws_locations
        try:
            # if we can import ZEO, assume the depenencies (zdaemon, ZConfig,
            # etc...) are also present on 'path'
            import ZEO
        except ImportError:
            raise AssertionError(zodb_import_msg)

        if os.path.exists(location):
            shutil.rmtree(location)

        # What follows is a bit of a hack because the instance-setup mechanism
        # is a bit monolithic. We'll run mkzeoinst and then we'll
        # patch the result. A better approach might be to provide independent
        # instance-creation logic, but this raises lots of issues that
        # need to be stored out first.

        # this was taken from mkzeoinstance.py
        from ZEO.mkzeoinst import ZEOInstanceBuilder

        self.zodb3_home = os.path.dirname(os.path.dirname(ZEO.__file__))
        params = {
            "package": "zeo",
            "PACKAGE": "ZEO",
            "zodb3_home": self.zodb3_home,
            "instance_home": location,
            "port": 8100, # will be overwritten later
            "python": options['executable'],
            }
        ZEOInstanceBuilder().create(location, params)

        try:
            # Save the working set:
            open(os.path.join(location, 'etc', '.eggs'), 'w').write(
                '\n'.join(self.ws_locations))

            # Make a new zeo.conf based on options in buildout.cfg
            self.build_zeo_conf()

            # Patch extra paths into binaries
            self.patch_binaries()

            # Install extra scripts
            self.install_scripts()

        except:
            # clean up
            shutil.rmtree(location)
            raise

        # end evil hack and restore sys.path to it's glorious days
        del sys.path[0:len(self.ws_locations)]
        assert tuple(sys.path) == orig_sys_path

        return location

    def update(self):
        options = self.options
        location = options['location']

        if os.path.exists(location):
            # See if we can stop. We need to see if the working set path
            # has changed.
            saved_path = os.path.join(location, 'etc', '.eggs')
            if os.path.isfile(saved_path):
                if (open(saved_path).read() !=
                    '\n'.join(self.ws_locations)
                    ):
                    # Something has changed. Blow away the instance.
                    self.install()

            # Nothing has changed.
            return location

        else:
            self.install()

        return location

    def build_zeo_conf(self):
        """Create a zeo.conf file
        """
        options = self.options
        location = options['location']
        instance_home = location

        zeo_conf_path = options.get('zeo-conf', None)
        if zeo_conf_path is not None:
            zeo_conf = "%%include %s" % os.path.abspath(zeo_conf_path)
        else:
            zeo_address = options.get('zeo-address', '8100')
            zeo_conf_additional = options.get('zeo-conf-additional', '')
            storage_number = options.get('storage-number', '1')

            monitor_address = options.get('monitor-address', '')
            if monitor_address:
                monitor_address = 'monitor-address %s' % monitor_address

            effective_user = options.get('effective-user', '')
            if effective_user:
               effective_user = 'user %s' % effective_user

            invalidation_queue_size = options.get('invalidation-queue-size',
                                                  '100')

            base_dir = self.buildout['buildout']['directory']
            socket_name = options.get('socket-name',
                                      '%s/var/zeo.zdsock' % base_dir)
            socket_dir = os.path.dirname(socket_name)
            if not os.path.exists(socket_dir):
                os.makedirs(socket_dir)

            z_log_name = os.path.sep.join(('var', 'log', self.name + '.log',))
            zeo_log_level = options.get('zeo-log-level', 'info')
            zeo_log_custom = options.get('zeo-log-custom', None)

            # if zeo-log is given, we use it to set the runner
            # logfile value in any case
            z_log_filename = options.get('zeo-log', z_log_name)
            z_log_filename = os.path.join(base_dir, z_log_filename)
            z_log_dir = os.path.dirname(z_log_filename)
            if not os.path.exists(z_log_dir):
                os.makedirs(z_log_dir)

            # zeo-log-custom superseeds zeo-log
            logformat=options.get('zeo-log-format', '%(asctime)s %(message)s')
            if zeo_log_custom is None:
                z_log = z_log_file % {'filename': z_log_filename, 'logformat' : logformat}
            else:
                z_log = zeo_log_custom

            file_storage = os.path.sep.join(('var', 'filestorage', 'Data.fs',))
            file_storage = options.get('file-storage', file_storage)
            file_storage = os.path.join(base_dir, file_storage)
            file_storage_dir = os.path.dirname(file_storage)
            if not os.path.exists(file_storage_dir):
                os.makedirs(file_storage_dir)

            if options.get('authentication-database', ''):
                authentication = authentication_template % \
                        dict(database=options.get('authentication-database'),
                             realm=options.get('authentication-realm', 'ZEO'))
            else:
                authentication = ''

            pid_file = options.get(
                'pid-file',
                os.path.join(base_dir, 'var', self.name + '.pid'))

            # check whether we'll wrap a blob storage around our file storage
            blob_storage = options.get('blob-storage', None)
            if blob_storage is not None:
                storage_template = blob_storage_template
            else:
                storage_template = file_storage_template

            storage = storage_template % dict(
                storage_number = storage_number,
                file_storage = file_storage,
                blob_storage = blob_storage
                )

            zeo_conf = zeo_conf_template % dict(
                instance_home = instance_home,
                effective_user = effective_user,
                invalidation_queue_size = invalidation_queue_size,
                socket_name = socket_name,
                z_log = z_log,
                z_log_filename = z_log_filename,
                authentication = authentication,
                storage = storage,
                zeo_address = zeo_address,
                pid_file = pid_file,
                zeo_conf_additional = zeo_conf_additional,
                monitor_address = monitor_address,
                zeo_log_level = zeo_log_level,
                )

        zeo_conf_path = os.path.join(location, 'etc', 'zeo.conf')
        open(zeo_conf_path, 'w').write(zeo_conf)

    def patch_binaries(self):
        location = self.options['location']
        # XXX We need to patch the windows specific batch scripts
        # and they need a different path seperator
        path = os.path.pathsep.join(self.ws_locations)
        for script_name in ('runzeo', 'zeoctl'):
            script_path = os.path.join(location, 'bin', script_name)
            script = open(script_path).read()
            script = script.replace('PYTHONPATH="$ZODB3_HOME"',
                                    'PYTHONPATH="%s"' % path)
            f = open(script_path, 'w')
            f.write(script)
            f.close()

    def install_scripts(self):
        options = self.options
        location = options['location']

        self.zeo_conf = options.get('zeo-conf', None)
        if self.zeo_conf is None:
            self.zeo_conf = os.path.join(location, 'etc', 'zeo.conf')

        _, ws = self.egg.working_set(['plone.recipe.zope2zeoserver'])

        initialization = """
        import os; os.environ['PYTHONPATH'] = %r
        """.strip() % os.path.pathsep.join(self.ws_locations)
        zc.buildout.easy_install.scripts(
            [(self.name, 'plone.recipe.zope2zeoserver.ctl', 'main')],
            ws, options['executable'], options['bin-directory'],
            extra_paths = self.ws_locations,
            initialization = initialization,
            arguments = ('\n        ["-C", %r]'
                         '\n        + sys.argv[1:]'
                         % self.zeo_conf
                         ),
            relative_paths=self._relative_paths,
            )
        # zeopack.py
        script_name = options.get('zeopack-script-name', 'zeopack')
        zeopack = options.get('zeopack', None)
        if zeopack is not None:
            directory, filename = os.path.split(zeopack)
            if zeopack and os.path.exists(zeopack):
                zc.buildout.easy_install.scripts(
                    [(script_name, os.path.splitext(filename)[0], 'main')],
                    {}, options['executable'], options['bin-directory'],
                    extra_paths = self.ws_locations + [directory],
                    relative_paths=self._relative_paths,
                    )
        else:
            host = port = socket_path = ''
            zeo_address = options.get('zeo-address', '8100')
            parts = zeo_address.split(':')

            if len(parts) == 1:
                try:
                    # if the only argument is a port, which must be an int,
                    # we use 127.0.0.1 as the host by default
                    int_port = int(zeo_address)
                    host = '127.0.0.1'
                    port = zeo_address
                except ValueError:
                    # The address is a simple string, we now assume it is
                    # a path to a Unix socket file
                    socket_path = zeo_address
            else:
                host, port = parts

            username = options.get('pack-user', None)
            password = options.get('pack-password', None)
            if username is not None:
                realm = options.get('authentication-realm', 'ZEO')
            else:
                realm = None

            arg_list = [
                'host', 'port', 'unix', 'days',
                'username', 'password', 'realm', 'blob_dir', 'storage',
            ]
            arguments = dict(
                host=host,
                port=port,
                unix=socket_path,
                days=options.get('pack-days', 1),
                username=username,
                password=password,
                realm=realm,
                blob_dir=options.get('blob-storage', None),
            )
            arguments_info = ''
            for k,v in arguments.items():
                if not v:
                    arguments_info += '%s = None\n' % k
                else:
                    arguments_info += '%s = "%s"\n' % (k, v)
            arguments_info += ("import getopt; opts = getopt.getopt("
                               "sys.argv[1:], 'S:W1')[0]; storage = "
                               "opts and opts[0][1] or '1'")
            extra_paths = [location for location in self.ws_locations
                           if location not in self.zodb_ws.entries]
            # Make sure the recipe itself and its dependencies are on the path
            extra_paths.append(ws.by_key[options['recipe']].location)
            extra_paths.append(ws.by_key['zc.buildout'].location)
            extra_paths.append(ws.by_key['zc.recipe.egg'].location)
            zc.buildout.easy_install.scripts(
                [(script_name, 'plone.recipe.zope2zeoserver.pack', 'main')],
                self.zodb_ws, options['executable'], options['bin-directory'],
                initialization=arguments_info,
                arguments=', '.join(arg_list), extra_paths=extra_paths,
                relative_paths=self._relative_paths,
                )

        # The backup script, pointing to repozo.py
        script_name = options.get('repozo-script-name', 'repozo')
        repozo = options.get('repozo', None)
        if repozo is None:
            repozo = os.path.join(options['zope2-location'], 'utilities', 'ZODBTools', 'repozo.py')

        directory, filename = os.path.split(repozo)
        if repozo and os.path.exists(repozo):
            zc.buildout.easy_install.scripts(
                [(script_name, os.path.splitext(filename)[0], 'main')],
                {}, options['executable'], options['bin-directory'],
                extra_paths = [os.path.join(options['zope2-location'], 'lib', 'python'),
                               directory],
                relative_paths=self._relative_paths,
                )

        if sys.platform == 'win32':
            self.install_win32_scripts()

    def install_win32_scripts(self):
            location = self.options['location']

            zope2_location = self.options.get('zope2-location', None)

            if zope2_location is not None:
                software_home = join(zope2_location, 'lib', 'python')
            else:
                software_home = None

            arguments = {'PYTHON': self.options['executable'],
                         'SOFTWARE_HOME': software_home,
                         'zodb3_home': self.zodb3_home,
                         'ZOPE_HOME': zope2_location,
                         'INSTANCE_HOME': location,
                         'PYTHONPATH': os.path.pathsep.join(self.ws_locations),
                         'PACKAGE': 'zeo'}

            if zope2_location is not None:
                # zeoservice.py
                # requires zope2_location due to the nt_svcutils package
                zeo_filename = '%s_service' % self.name
                zeo_service = open(join(curdir, 'zeoservice.py.in')).read()
                zeo_file = os.path.join(self.options['bin-directory'],
                                        '%s.py' % zeo_filename)
                self._write_file(zeo_file, zeo_service % arguments)

                initialization = """
                import os; os.environ['PYTHONPATH'] = %r
                """.strip() % os.path.pathsep.join(self.ws_locations)

                zc.buildout.easy_install.scripts(
                    [(zeo_filename, zeo_filename, 'main')],
                    self.zodb_ws,
                    self.options['executable'],
                    self.options['bin-directory'],
                    extra_paths = self.ws_locations,
                    initialization = initialization,
                    )


            # runzeo.bat
            runzeo_filename = '%s_runzeo.bat' % self.name
            runzeo = open(join(curdir, 'runzeo.bat')).read()
            self._write_file(os.path.join(self.options['bin-directory'],
                             runzeo_filename), runzeo % arguments)


    def _write_file(self, path, content):
        logger = logging.getLogger('zc.buildout.easy_install')
        f = open(path, 'w')
        try:
            f.write(content)
        finally:
            f.close()
        logger.debug('Wrote file %s' % path)
        os.chmod(path, 0755)
        logger.warning('Changed mode for %s to 755' % path)


invalid_z2_location_msg = """
'zope2-location' (%r) doesn't point to an actual Zope2 SOFTWARE_HOME. Check this
setting and, eventually, make sure this buildout part is being run after the
part that is supposed to provide the 'zope2-location' value.
""".strip()

zodb_import_msg = """
Unable to import ZEO. Please, either add the ZODB3 egg to the 'eggs' entry or
set zope2-location
""".strip()

# the template used to build a regular file storage entry for zeo.conf
file_storage_template = """
<filestorage %(storage_number)s>
  path %(file_storage)s
</filestorage>
""".strip()

# the template used to build a blob storage wrapping a file storage entry for
# zeo.conf
blob_storage_template = """
<blobstorage %(storage_number)s>
  blob-dir %(blob_storage)s
  <filestorage %(storage_number)s>
    path %(file_storage)s
  </filestorage>
</blobstorage>
""".strip()

z_log_file = """\
     <logfile>
      path %(filename)s
      format %(logformat)s
    </logfile>
""".strip()

authentication_template = """
  authentication-protocol digest
  authentication-database %(database)s
  authentication-realm %(realm)s
"""

# The template used to build zeo.conf
zeo_conf_template = """\
%%define INSTANCE %(instance_home)s

<zeo>
  address %(zeo_address)s
  read-only false
  invalidation-queue-size %(invalidation_queue_size)s
  pid-filename %(pid_file)s
  %(authentication)s
  %(monitor_address)s
</zeo>

%(storage)s

<eventlog>
  level %(zeo_log_level)s
  %(z_log)s
</eventlog>

<runner>
  program $INSTANCE/bin/runzeo
  socket-name %(socket_name)s
  daemon true
  forever false
  backoff-limit 10
  exit-codes 0, 2
  directory $INSTANCE
  default-to-interactive true
  %(effective_user)s

  # This logfile should match the one in the zeo.conf file.
  # It is used by zdctl's logtail command, zdrun/zdctl doesn't write it.
  logfile %(z_log_filename)s
</runner>

%(zeo_conf_additional)s
"""
